Add client_id and id_token_hint to IdP logout

ubipo opened the pull request:
_Edited to add links to docs/code examples_

This adds two parameters to the end_session_endpoint IdP URL which the
user gets redirected to when singleLogout is triggered.

These paramters are:
- client_id: the client ID of the current session's provider. 'OPTIONAL'
  as per the relevant OpenID specification.
- id_token_hint: the raw id_token that was obtained during the code
  callback of this session's login flow (set in session variable `oidc.id_token`). 'RECOMMENDED' by the relevant OpenID specification [1].

Some providers (e.g. node-oidc-provider [2] and Keycloak [3]) require this when using the code OAuth flow.

Because passing id_token_hint reveals the id_token to the user agent, a
app setting was also added to optionally turn this behaviour off (default is
turned on).

Builds upon PR https://github.com/nextcloud/user_oidc/pull/373 / issue https://github.com/nextcloud/user_oidc/issues/336
Fixes issue https://github.com/nextcloud/user_oidc/issues/449

[1]: https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RPLogout
[2]: https://github.com/panva/node-oidc-provider/blob/c243bf6b6663c41ff3e75c09b95fb978eba87381/lib/actions/end_session.js#L32
[3]: https://www.keycloak.org/docs/latest/release_notes/index.html#oidc-logout-changes

Signed-off-by: Pieter Fiers <pieter@pfiers.net>

eneiluj commented on lib/Service/ProviderService.php:
```diff
@@ -43,6 +44,10 @@ class ProviderService {
 	public const SETTING_EXTRA_CLAIMS = 'extraClaims';
 	public const SETTING_JWKS_CACHE = 'jwksCache';
 	public const SETTING_JWKS_CACHE_TIMESTAMP = 'jwksCacheTimestamp';
+	private const BOOLEAN_SETTINGS = array(
+		self::SETTING_UNIQUE_UID, self::SETTING_CHECK_BEARER,
+		self::SETTING_SEND_ID_TOKEN_HINT
+	);
```
```suggestion
	private const BOOLEAN_SETTINGS = [
		self::SETTING_UNIQUE_UID,
		self::SETTING_CHECK_BEARER,
		self::SETTING_SEND_ID_TOKEN_HINT,
	];
```

eneiluj commented on src/components/SettingsForm.vue:
```diff
@@ -108,6 +108,12 @@
+			{{ t('user_oidc', 'Send ID token hint on logout') }}
+		</CheckboxRadioSwitch>
+		<p class="settings-hint">
+			{{ t('user_oidc', 'Should the ID token be included as a parameter (id_token_hint) of the provider\'s logout URL? The user will be redirected to this URL after logout in Nextcloud. Enabling this exposes the ID token to the user agent, which may not be necessary depending on the particular OIDC provider used.') }}
```
```suggestion
			{{ t('user_oidc', 'Should the ID token be included as a the id_token_hint GET parameter in the OpenID logout URL? User are redirected to this URL after logging out of Nextcloud. Enabling this setting exposes the OIDC ID token to the user agent, which may not be necessary depending on the OIDC provider.') }}
```

eneiluj reviewed (changes requested):
Nice! Thanks for the PR. Nicely done.

I had trouble finding doc about why the client_id is needed but here it is: https://www.keycloak.org/docs/latest/release_notes/index.html#oidc-logout-changes

Could you sign your commits and rebase your branch on this project's master?

ubipo commented:
Thanks for the review! I've implemented the suggestions, rebased onto master, signed the commit, and added some more info on the use of the parameters. Should be good to go :)

eneiluj reviewed (approved):
:+1:
